server/src/test/java/io/druid/server/coordinator/CuratorDruidCoordinatorTest.java

```diff
@@ -0,0 +1,522 @@
+
+    // temporarily set latch count to 1
+    segmentRemovedLatch = new CountDownLatch(0);
+
+    setupView();
+
+    DruidServer source = new DruidServer(
+        "localhost:1",
+        "localhost:1",
+        null,
+        10000000L,
+        ServerType.HISTORICAL,
+        "default_tier",
+        0
+    );
+
+    DruidServer dest = new DruidServer(
+        "localhost:2",
+        "localhost:2",
+        null,
+        10000000L,
+        ServerType.HISTORICAL,
+        "default_tier",
+        0
+    );
+
+    setupZNodeForServer(source, zkPathsConfig, jsonMapper);
+    setupZNodeForServer(dest, zkPathsConfig, jsonMapper);
+
+    final List<DataSegment> sourceSegments = Lists.transform(
+        ImmutableList.of(
+            Pair.of("2011-04-01/2011-04-03", "v1"),
+            Pair.of("2011-04-03/2011-04-06", "v1"),
+            Pair.of("2011-04-06/2011-04-09", "v1")
+        ),
+        input -> dataSegmentWithIntervalAndVersion(input.lhs, input.rhs)
+    );
+
+    final List<DataSegment> destinationSegments = Lists.transform(
+        ImmutableList.of(
+            Pair.of("2011-03-31/2011-04-01", "v1")
+        ),
+        input -> dataSegmentWithIntervalAndVersion(input.lhs, input.rhs)
+    );
+
+    DataSegment segmentToMove = sourceSegments.get(2);
+
+    List<String> sourceSegKeys = Lists.newArrayList();
+    List<String> destSegKeys = Lists.newArrayList();
+
+    for (DataSegment segment : sourceSegments) {
+      sourceSegKeys.add(announceBatchSegmentsForServer(source, ImmutableSet.of(segment), zkPathsConfig, jsonMapper));
+    }
+
+    for (DataSegment segment : destinationSegments) {
+      destSegKeys.add(announceBatchSegmentsForServer(dest, ImmutableSet.of(segment), zkPathsConfig, jsonMapper));
+    }
+
+    Assert.assertTrue(timing.forWaiting().awaitLatch(segmentViewInitLatch));
+    Assert.assertTrue(timing.forWaiting().awaitLatch(segmentAddedLatch));
+
+    sourceLoadQueueChildrenCache.getListenable().addListener(
+        (curatorFramework, pathChildrenCacheEvent) -> {
+          if (pathChildrenCacheEvent.getType().equals(PathChildrenCacheEvent.Type.CHILD_ADDED)) {
+            //Simulate source server dropping segment
+            unannounceSegmentFromBatchForServer(source, segmentToMove, sourceSegKeys.get(2), zkPathsConfig);
+            segmentRemovedLatch.countDown();
+          }
+        }
+    );
+
+    destinationLoadQueueChildrenCache.getListenable().addListener(
+        (curatorFramework, pathChildrenCacheEvent) -> {
+          if (pathChildrenCacheEvent.getType().equals(PathChildrenCacheEvent.Type.CHILD_ADDED)) {
+            //Simulate destination server loading segment
+            announceBatchSegmentsForServer(dest, ImmutableSet.of(segmentToMove), zkPathsConfig, jsonMapper);
+            segmentAddedLatch.countDown();
+          }
+        }
+    );
+
+    sourceLoadQueueChildrenCache.start();
+    destinationLoadQueueChildrenCache.start();
+
+    Thread.sleep(100);
```

Comment:
This might cause test failures in Travis. If this is to wait for childrenCaches to be ready, some kind of real check (by polling or using a latch) should be used instead. Same for other sleeps.